Legend Component

alancleary opened the pull request:
Looking for feedback and testing.

Note that not all features mentioned in [the spec](https://github.com/legumeinfo/website-ui-specs/tree/main/legend-component) aren't currently implemented. Specifically, only mouse click is implemented (no mouse over/out) and there is no control over the position of the color glyph, although now that I mention this it is seems like a property to control this could subsume the `compact` property. I didn't implement these bits because I don't see an immediate use case for them.

Also note that currently only entry colors can be overridden on a per-entry basis, i.e. if an entry has a `color` attribute then this will be used instead of a color assigned by the `colorFunction`. I could similar overriding functionality being added for the `glyph` and `compact` properties, although I think there are fewer use cases for this.

adf-ncgr commented:
thinking about legend entry order a little more, I guess this is probably something that shouldn't be the dumb component's job (e.g. if the color scheme were like what we use for gene families in gcv, sorting by color would make no sense). But maybe including an orderingFunction just as we have a colorFunction would be the way to address this (not sure, just throwing it out there)

adf-ncgr commented:
OK, I can buy that argument, although it seems like the same argument could apply to colors (ie apply both functions externally and pass in an ordered map of color assignments to the component). But I'm not clear on what the internal logic is that you referred to above to justify including colorFunction.

StevenCannon-USDA commented:
It looks good to me.

Regarding ordering, this seems sensible to me: "draw the elements in the order given." 

Glancing at the code, it looks like the elements are passed in using an array?
 ```
 *   const data = {
 *     entries: [
 *       {label: 'Acacia crassicarpa'},
 *       {label: 'Aeschynomene evenia'},
 *       {label: 'Apios americana'},
 *     ],
 *   };
 ```

Some inconsequential feedback re configuration: in the phylotree context, I think I would choose 
  Horizontal / Circle / not-Compact
(not-Compact for text readability, and I like the dots in the legend matching the dots in the tree)

Intro text: typo "Sybology"

alancleary commented:
> it looks like the elements are passed in using an array?

Yes. The data is given as an object with the elements as an `entries` attribute of the object. I did this in case we want to add more/different data in the future, i.e. more data can be added as new object attributes. Honestly, this is probably a little overengineered and the input data should simply be the array of elements.

> I think I would choose Horizontal / Circle / not-Compact

I'll change this.

> I like the dots in the legend matching the dots in the tree

Can you be more specific about this? Should they be the same size? Should they also be outlined?

StevenCannon-USDA commented:
> Can you be more specific about this? Should they be the same size? Should they also be outlined?

I was about to say that it would probably be best for the symbols to be the same (same size, and outlined), since that is typical for a legend+diagram -- but on second thought, I like how they appear in the legend component as it is (dots larger and without outline), since the legend text is larger than the text in the phylotree; larger text, larger symbols. So, my vote: the legend dots are fine!

<img width="399" alt="legend_dots" src="https://github.com/user-attachments/assets/2609f0f9-7e84-4937-aba8-85696eac105c" /> <img width="187" alt="phylotree" src="https://github.com/user-attachments/assets/dcd10a53-ba30-4db2-92bb-d8821f031f01" />

alancleary commented:
@StevenCannon-USDA @adf-ncgr I updated the legend component to consume an array instead of an object and introduced a new `position` property that sets the position of the glyphs relative to their labels. This new property subsumed the `compact` property with its `contain` option. A new `right` option was added to this property, which takes us a step closer to supporting how GCV currently draws its legend.

Let me know if you think there's any other changes that need to be made to satisfy the spec. If not, I'll go ahead and mark this draft as ready for review.

StevenCannon-USDA commented:
@alancleary For some reason, after the changes I am not seeing the legend - either at http://localhost:8000/dev/examples/lis-phylotree-with-linkouts.html or at http://localhost:8000/dev/lis-legend-element.html. I have tried clearing cache and viewing in an incognito window.

alancleary commented:
> @alancleary For some reason, after the changes I am not seeing the legend - either at http://localhost:8000/dev/examples/lis-phylotree-with-linkouts.html or at http://localhost:8000/dev/lis-legend-element.html. I have tried clearing cache and viewing in an incognito window.

Are there any errors in the web console? Also note that you'll need to build the library after pulling the changes.

StevenCannon-USDA commented:
Ah, I had forgotten to build.
The changes all look good to me.

alancleary commented:
OK. I've marked the PR as ready for review. @StevenCannon-USDA has already approved it. @adf-ncgr let me know if you want to take a final look or if you want @ctcncgr to evaluate it before merging.

adf-ncgr reviewed (approved):
looks good- ship it!
